File: Source/Client/Game/Objects/Projectile.vb
Imports System.IO
Imports Mirage.Sharp.Asfw
Imports Core

Friend Module Projectile
    
#Region "Sending"
    Sub SendRequestEditProjectile()
        Dim buffer As ByteStream

        buffer = New ByteStream(4)
        buffer.WriteInt32(Packets.ClientPackets.CRequestEditProjectiles)
        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()

    End Sub

    Sub SendSaveProjectile(ProjectileNum As Integer)
        Dim buffer As ByteStream

        buffer = New ByteStream(4)

        buffer.WriteInt32(ClientPackets.CSaveProjectile)
        buffer.WriteInt32(ProjectileNum)

        buffer.WriteString(Type.Projectile(ProjectileNum).Name)
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Sprite)
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Range)
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Speed)
        buffer.WriteInt32(Type.Projectile(ProjectileNum).Damage)

        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()

    End Sub

    Sub SendRequestProjectile()
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(ClientPackets.CRequestProjectiles)

        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()

    End Sub

    Sub SendClearProjectile(projectileNum As Integer, collisionindex As Integer, collisionType As Byte, collisionZone As Integer)
        Dim buffer As New ByteStream(4)

        buffer.WriteInt32(ClientPackets.CClearProjectile)
        buffer.WriteInt32(projectileNum)
        buffer.WriteInt32(collisionindex)
        buffer.WriteInt32(collisionType)
        buffer.WriteInt32(collisionZone)

        Socket.SendData(buffer.Data, buffer.Head)
        buffer.Dispose()

    End Sub

#End Region

#Region "Recieving"

    Friend Sub HandleUpdateProjectile(ByRef data() As Byte)
        Dim projectileNum As Integer
        Dim buffer As New ByteStream(data)
        projectileNum = buffer.ReadInt32

        Type.Projectile(projectileNum).Name = buffer.ReadString
        Type.Projectile(projectileNum).Sprite = buffer.ReadInt32
        Type.Projectile(projectileNum).Range = buffer.ReadInt32
        Type.Projectile(projectileNum).Speed = buffer.ReadInt32
        Type.Projectile(projectileNum).Damage = buffer.ReadInt32

        buffer.Dispose()

    End Sub

    Friend Sub HandleMapProjectile(ByRef data() As Byte)
        Dim i As Integer
        Dim buffer As New ByteStream(data)
        i = buffer.ReadInt32

        With MapProjectile(Type.Player(GameState.MyIndex).Map, i)
            .ProjectileNum = buffer.ReadInt32
            .Owner = buffer.ReadInt32
            .OwnerType = buffer.ReadInt32
            .Dir = buffer.ReadInt32
            .X = buffer.ReadInt32
            .Y = buffer.ReadInt32
            .Range = 0
            .Timer = GetTickCount() + 60000
        End With

        buffer.Dispose()

    End Sub

#End Region

#Region "Database"

    Sub ClearProjectile()
        Dim i As Integer

       For i = 1 To MAX_PROJECTILES
            ClearProjectile(i)
        Next

    End Sub

    Sub ClearProjectile(index As Integer)

        Type.Projectile(index).Name = ""
        Type.Projectile(index).Sprite = 0
        Type.Projectile(index).Range = 0
        Type.Projectile(index).Speed = 0
        Type.Projectile(index).Damage = 0

    End Sub

    Sub ClearMapProjectile(projectileNum As Integer)

        MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).ProjectileNum = 0
        MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Owner = 0
        MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).OwnerType = 0
        MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).X = 0
        MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Y = 0
        MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Dir = 0
        MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Timer = 0

    End Sub

#End Region

#Region "Drawing"
    Friend Sub DrawProjectile(projectileNum As Integer)
        Dim rec As RectStruct
        Dim canClearProjectile As Boolean
        Dim collisionindex As Integer
        Dim collisionType As Byte
        Dim collisionZone As Integer
        Dim xOffset As Integer, yOffset As Integer
        Dim x As Integer, y As Integer
        Dim i As Integer
        Dim sprite As Integer

        ' check to see if it's time to move the Projectile
        If GetTickCount() > MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).TravelTime Then
            Select Case MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Dir
                Case DirectionType.Up
                    MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Y = MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Y - 1
                Case DirectionType.Down
                    MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Y = MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Y + 1
                Case DirectionType.Left
                    MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).X = MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).X - 1
                Case DirectionType.Right
                    MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).X = MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).X + 1
            End Select
            MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).TravelTime = GetTickCount() + Type.Projectile(Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).ProjectileNum).Speed
            MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Range = MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Range + 1
        End If

        x = MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).X
        y = MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Y

        'Check if its been going for over 1 minute, if so clear.
       If Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Timer < GetTickCount() Then canClearProjectile = 1

        If x > MyMap.MaxX Or x < 0 Then canClearProjectile = 1
        If y > MyMap.MaxY Or y < 0 Then canClearProjectile = 1

        'Check for blocked wall collision
        If canClearProjectile = 0 Then 'Add a check to prevent crashing
            If MyMap.Tile(x, y).Type = TileType.Blocked Or MyMap.Tile(x, y).Type2 = TileType.Blocked Then
                canClearProjectile = 1
            End If
        End If

        'Check for npc collision
       For i = 1 To MAX_MAP_NPCS
            If MyMapNPC(i).X = x And MyMapNPC(i).Y = y Then
                canClearProjectile = 1
                collisionindex = i
                collisionType = TargetType.NPC
                collisionZone = -1
                Exit For
            End If
        Next

        'Check for player collision
       For i = 1 To MAX_PLAYERS
            If IsPlaying(i) And GetPlayerMap(i) = GetPlayerMap(GameState.MyIndex) Then
                If GetPlayerX(i) = x And GetPlayerY(i) = y Then
                    canClearProjectile = 1
                    collisionindex = i
                    collisionType = TargetType.Player
                    collisionZone = -1
                   If Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).OwnerType = TargetType.Player Then
                       If Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Owner = i Then canClearProjectile = 0 ' Reset if its the owner of projectile
                    End If
                    Exit For
                End If

            End If
        Next

        'Check if it has hit its maximum range
       If Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Range >= Type.Projectile(Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).ProjectileNum).Range + 1 Then canClearProjectile = 1

        'Clear the projectile if possible
        If canClearProjectile = 1 Then
            'Only send the clear to the server if you're the projectile caster or the one hit (only if owner is not a player)
            If (Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).OwnerType = TargetType.Player And MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Owner = GameState.MyIndex) Then
                SendClearProjectile(projectileNum, collisionindex, collisionType, collisionZone)
            End If

            ClearMapProjectile(projectileNum)
            Exit Sub
        End If

        sprite = Type.Projectile(Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).ProjectileNum).Sprite
        If sprite < 1 Or sprite > GameState.NumProjectiles Then Exit Sub

        ' src rect
        With rec
            .Top = 0
            .Bottom = GameClient.GetGfxInfo(System.IO.Path.Combine(Core.Path.Projectiles, sprite)).Height
            .Left = MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Dir * GameState.PicX
            .Right = .Left + GameState.PicX
        End With

        'Find the offset
        Select Case MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).Dir
            Case DirectionType.Up
                yOffset = ((Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).TravelTime - GetTickCount()) / Type.Projectile(Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).ProjectileNum).Speed) * GameState.PicY
            Case DirectionType.Down
                yOffset = -((Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).TravelTime - GetTickCount()) / Type.Projectile(Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).ProjectileNum).Speed) * GameState.PicY
            Case DirectionType.Left
                xOffset = ((Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).TravelTime - GetTickCount()) / Type.Projectile(Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).ProjectileNum).Speed) * GameState.PicX
            Case DirectionType.Right
                xOffset = -((Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).TravelTime - GetTickCount()) / Type.Projectile(Type.MapProjectile(Type.Player(GameState.MyIndex).Map, projectileNum).ProjectileNum).Speed) * GameState.PicX
        End Select

        ' Convert coordinates
        x = ConvertMapX(x * GameState.PicX)
        y = ConvertMapY(y * GameState.PicY)

        ' Render texture
        GameClient.RenderTexture(System.IO.Path.Combine(Core.Path.Projectiles & sprite), x, y, rec.Left, rec.Top, 32, 32)

    End Sub
    
#End Region

End Module

File: Source/Client/Game/Objects/Projectile.cs
using System;
using Core;
using static Core.Global.Command;
using Microsoft.VisualBasic.CompilerServices;
using Mirage.Sharp.Asfw;

namespace Client
{

    public class Projectile
    {

        #region Sending
        public static void SendRequestEditProjectiles()
        {
            ByteStream buffer;

            buffer = new ByteStream(4);
            buffer.WriteInt32((int)Packets.ClientPackets.CRequestEditProjectile);
            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();

        }

        public static void SendSaveProjectile(int ProjectileNum)
        {
            ByteStream buffer;

            buffer = new ByteStream(4);

            buffer.WriteInt32((int)Packets.ClientPackets.CSaveProjectile);
            buffer.WriteInt32(ProjectileNum);

            buffer.WriteString(Data.Projectile[ProjectileNum].Name);
            buffer.WriteInt32(Core.Data.Projectile[ProjectileNum].Sprite);
            buffer.WriteInt32(Core.Data.Projectile[ProjectileNum].Range);
            buffer.WriteInt32(Core.Data.Projectile[ProjectileNum].Speed);
            buffer.WriteInt32(Core.Data.Projectile[ProjectileNum].Damage);

            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();

        }

        public static void SendRequestProjectile(int projectileNum)
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32((int)Packets.ClientPackets.CRequestProjectile);
            buffer.WriteInt32(projectileNum);

            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();

        }

        public static void SendClearProjectile(int projectileNum, int collisionindex, byte collisionType, int collisionZone)
        {
            var buffer = new ByteStream(4);

            buffer.WriteInt32((int)Packets.ClientPackets.CClearProjectile);
            buffer.WriteInt32(projectileNum);
            buffer.WriteInt32(collisionindex);
            buffer.WriteInt32(collisionType);
            buffer.WriteInt32(collisionZone);

            NetworkConfig.Socket.SendData(buffer.UnreadData, buffer.WritePosition);
            buffer.Dispose();

        }

        #endregion

        #region Recieving

        public static void HandleUpdateProjectile(ref byte[] data)
        {
            int projectileNum;
            var buffer = new ByteStream(data);
            projectileNum = buffer.ReadInt32();

            Core.Data.Projectile[projectileNum].Name = buffer.ReadString();
            Core.Data.Projectile[projectileNum].Sprite = buffer.ReadInt32();
            Core.Data.Projectile[projectileNum].Range = (byte)buffer.ReadInt32();
            Core.Data.Projectile[projectileNum].Speed = buffer.ReadInt32();
            Core.Data.Projectile[projectileNum].Damage = buffer.ReadInt32();

            buffer.Dispose();

        }

        public static void HandleMapProjectile(ref byte[] data)
        {
            int i;
            var buffer = new ByteStream(data);
            i = buffer.ReadInt32();

            {
                ref var withBlock = ref Core.Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, i];
                withBlock.ProjectileNum = buffer.ReadInt32();
                withBlock.Owner = buffer.ReadInt32();
                withBlock.OwnerType = (byte)buffer.ReadInt32();
                withBlock.Dir = (byte)buffer.ReadInt32();
                withBlock.X = buffer.ReadInt32();
                withBlock.Y = buffer.ReadInt32();
                withBlock.Range = 0;
                withBlock.Timer = General.GetTickCount() + 60000;
            }

            buffer.Dispose();

        }

        #endregion

        #region Database

        public static void ClearProjectile()
        {
            int i;

            for (i = 0; i < Constant.MAX_PROJECTILES;  i++)
                ClearProjectile(i);

        }

        public static void ClearProjectile(int index)
        {
            Core.Data.Projectile[index].Name = "";
            Core.Data.Projectile[index].Sprite = 0;
            Core.Data.Projectile[index].Range = 0;
            Core.Data.Projectile[index].Speed = 0;
            Core.Data.Projectile[index].Damage = 0;

        }

        public static void ClearMapProjectile(int projectileNum)
        {
            Core.Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].ProjectileNum = 0;
            Core.Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Owner = 0;
            Core.Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].OwnerType = 0;
            Core.Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].X = 0;
            Core.Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Y = 0;
            Core.Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Dir = 0;
            Core.Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Timer = 0;

        }

        public static void StreamProjectile(int projectileNum)
        {
            if (projectileNum >= 0 & string.IsNullOrEmpty(Core.Data.Projectile[projectileNum].Name) && GameState.Projectile_Loaded[projectileNum] == 0)
            {
                GameState.Projectile_Loaded[projectileNum] = 1;
                SendRequestProjectile(projectileNum);
            }
        }

        #endregion

        #region Drawing
        public static void DrawProjectile(int projectileNum)
        {
            Core.Type.Rect rec;
            var canClearProjectile = default(bool);
            var collisionindex = default(int);
            var collisionType = default(byte);
            var collisionZone = default(int);
            int xOffset;
            int yOffset;
            int x;
            int y;
            int i;
            int sprite;

            StreamProjectile(projectileNum);

            // check to see if it's time to move the Projectile
            if (General.GetTickCount() > Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].TravelTime)
            {
                switch (Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Dir)
                {
                    case (byte)Direction.Up:
                        {
                            Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Y = Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Y - 1;
                            break;
                        }
                    case (byte)Direction.Down:
                        {
                            Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Y = Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Y + 1;
                            break;
                        }
                    case (byte)Direction.Left:
                        {
                            Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].X = Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].X - 1;
                            break;
                        }
                    case (byte)Direction.Right:
                        {
                            Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].X = Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].X + 1;
                            break;
                        }
                }
                Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].TravelTime = General.GetTickCount() + Data.Projectile[Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].ProjectileNum].Speed;
                Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Range = Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Range + 1;
            }

            x = Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].X;
            y = Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Y;

            // Check if its been going for over 1 minute, if so clear.
            if (Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Timer < General.GetTickCount())
                canClearProjectile = true;

            if (x > Data.MyMap.MaxX | x < 0)
                canClearProjectile = true;

            if (y > Data.MyMap.MaxY | y < 0)
                canClearProjectile = true;

            // Check for blocked wall collision
            if (Conversions.ToInteger(canClearProjectile) == 0) // Add a check to prevent crashing
            {
                if (Data.MyMap.Tile[x, y].Type == TileType.Blocked | Data.MyMap.Tile[x, y].Type2 == TileType.Blocked)
                {
                    canClearProjectile = true;
                }
            }

            // Check for Npc collision
            for (i = 0; i < Constant.MAX_MAP_NPCS; i++)
            {
                if (Data.MyMapNpc[i].X == x & Data.MyMapNpc[i].Y == y)
                {
                    canClearProjectile = true;
                    collisionindex = i;
                    collisionType = (byte)TargetType.Npc;
                    collisionZone = -1;
                    break;
                }
            }

            // Check for player collision
            for (i = 0; i < Constant.MAX_PLAYERS; i++)
            {
                if (IsPlaying(i) & GetPlayerMap(i) == GetPlayerMap(GameState.MyIndex))
                {
                    if (GetPlayerX(i) == x & GetPlayerY(i) == y)
                    {
                        canClearProjectile = true;
                        collisionindex = i;
                        collisionType = (byte)TargetType.Player;
                        collisionZone = -1;
                        if (Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].OwnerType == (byte)TargetType.Player)
                        {
                            if (Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Owner == i)
                                canClearProjectile = false; // Reset if its the owner of projectile
                        }
                        break;
                    }

                }
            }

            // Check if it has hit its maximum range
            if (Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Range >= Data.Projectile[Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].ProjectileNum].Range + 1)
                canClearProjectile = true;

            // Clear the projectile if possible
            if (Conversions.ToInteger(canClearProjectile) == 1)
            {
                // Only send the clear to the server if you're the projectile caster or the one hit (only if owner is not a player)
                if (Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].OwnerType == (byte)TargetType.Player & Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Owner == GameState.MyIndex)
                {
                    SendClearProjectile(projectileNum, collisionindex, collisionType, collisionZone);
                }

                ClearMapProjectile(projectileNum);
                return;
            }

            sprite = Data.Projectile[Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].ProjectileNum].Sprite;
            if (sprite < 1 | sprite > GameState.NumProjectiles)
                return;

            // src rect
            rec.Top = 0d;
            var gfxInfo = GameClient.GetGfxInfo(System.IO.Path.Combine(Core.Path.Projectiles, sprite.ToString()));
            if (gfxInfo == null)
            {
                return;
            }
            rec.Bottom = gfxInfo.Height;
            rec.Left = Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Dir * GameState.PicX;
            rec.Right = rec.Left + GameState.PicX;

            // Find the offset
            switch (Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].Dir)
            {
                case (byte)Direction.Up:
                    {
                        yOffset = (int)Math.Round((Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].TravelTime - General.GetTickCount()) / (double)Data.Projectile[Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].ProjectileNum].Speed * GameState.PicY);
                        break;
                    }
                case (byte)Direction.Down:
                    {
                        yOffset = (int)Math.Round(-((Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].TravelTime - General.GetTickCount()) / (double)Data.Projectile[Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].ProjectileNum].Speed) * GameState.PicY);
                        break;
                    }
                case (byte)Direction.Left:
                    {
                        xOffset = (int)Math.Round((Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].TravelTime - General.GetTickCount()) / (double)Data.Projectile[Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].ProjectileNum].Speed * GameState.PicX);
                        break;
                    }
                case (byte)Direction.Right:
                    {
                        xOffset = (int)Math.Round(-((Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].TravelTime - General.GetTickCount()) / (double)Data.Projectile[Data.MapProjectile[Core.Data.Player[GameState.MyIndex].Map, projectileNum].ProjectileNum].Speed) * GameState.PicX);
                        break;
                    }
            }

            // Convert coordinates
            x = GameLogic.ConvertMapX(x * GameState.PicX);
            y = GameLogic.ConvertMapY(y * GameState.PicY);

            // Render texture
            string argpath = System.IO.Path.Combine(Core.Path.Projectiles, sprite.ToString());
            GameClient.RenderTexture(ref argpath, x, y, (int)Math.Round(rec.Left), (int)Math.Round(rec.Top), 32, 32);

        }

        #endregion

    }
}